pkg/rehearse/jobs.go

```diff
@@ -188,18 +201,70 @@ func ExecuteJobs(toBeRehearsed map[string][]prowconfig.Presubmit, prNumber int,
 		}
 	}
 
-	if len(rehearsals) > 0 {
-		for _, job := range rehearsals {
-			created, err := submitRehearsal(job, refs, logger, pjclient)
-			if err != nil {
-				logger.WithError(err).Warn("Failed to execute a rehearsal presubmit")
-			} else {
-				logger.WithFields(pjutil.ProwJobFields(created)).Info("Submitted rehearsal prowjob")
-			}
-		}
-	} else {
+	if len(rehearsals) == 0 {
 		logger.Warn("No job rehearsals")
+		return nil
+	}
+	pjs := make(sets.String, len(rehearsals))
+	for _, job := range rehearsals {
+		created, err := submitRehearsal(job, refs, logger, pjclient)
+		if err != nil {
+			logger.WithError(err).Warn("Failed to execute a rehearsal presubmit")
+			continue
+		}
+		logger.WithFields(pjutil.ProwJobFields(created)).Info("Submitted rehearsal prowjob")
+		pjs.Insert(created.Name)
+	}
+	req, err := labels.NewRequirement(rehearseLabel, selection.Equals, []string{strconv.Itoa(prNumber)})
+	if err != nil {
+		return fmt.Errorf("failed to create label selector: %v", err)
+	}
+	selector := labels.NewSelector().Add(*req).String()
+	success, err := waitForJobs(pjs, selector, pjclient, logger)
+	if err != nil {
+		return fmt.Errorf("error while waiting for jobs: %s", err)
+	}
+	if !success {
+		return fmt.Errorf("not all ProwJobs were successful")
 	}
-
 	return nil
 }
+
+func waitForJobs(jobs sets.String, selector string, pjclient pj.ProwJobInterface, logger logrus.FieldLogger) (bool, error) {
+	if len(jobs) == 0 {
+		return true, nil
+	}
+	success := true
+	for {
+		w, err := pjclient.Watch(metav1.ListOptions{LabelSelector: selector})
+		if err != nil {
+			return false, fmt.Errorf("failed to create watch for ProwJobs: %v", err)
+		}
+		defer w.Stop()
+		ch := w.ResultChan()
+		for {
+			event, ok := <-ch
+			if !ok {
+				break
+			}
+			pj, ok := event.Object.(*pjapi.ProwJob)
+			if !ok {
+				return false, fmt.Errorf("received a %T from watch", event.Object)
+			}
+			logger.WithFields(pjutil.ProwJobFields(pj)).WithFields(logrus.Fields{"state": pj.Status.State}).Info("processing ProwJob")
```

Comment:
nit: the second `WithFields` can be a simpler https://godoc.org/github.com/sirupsen/logrus#WithField

Reply:
Nice, updated.